pandas/_libs/src/ujson/python/objToJSON.c

```diff
@@ -1512,8 +1532,17 @@ void Object_beginTypeContext(JSOBJ _obj, JSONTypeContext *tc) {
         tc->type = JT_UTF8;
         return;
     } else if (object_is_decimal_type(obj)) {
-        GET_TC(tc)->doubleValue = PyFloat_AsDouble(obj);
-        tc->type = JT_DOUBLE;
+        /* Check for null, since null can't go thru double path */
+        PyObject *is_null_obj = PyObject_CallMethod(obj,
+                                                    "is_nan",
+                                                    NULL);
+        if (is_null_obj == Py_False) {
+            GET_TC(tc)->doubleValue = PyFloat_AsDouble(obj);
+            tc->type = JT_DOUBLE;
+        } else {
+            tc->type = JT_NULL;
+        }
+        Py_DECREF(is_null_obj);
```

Comment:
In a case where `PyObject_CallMethod` returns `NULL` this would try to Py_DECREF `NULL`